crates/server/src/middleware/host.rs

```diff
@@ -0,0 +1,146 @@
+use std::convert::Infallible;
+use std::task::{Context, Poll};
+
+use axum::body::Body;
+use axum::extract::Request;
+use axum::http::{HeaderMap, StatusCode};
+use axum::response::{IntoResponse, Response};
+use libp2p::futures::future::BoxFuture;
+use multiaddr::Multiaddr;
+use tower::{Layer, Service};
+
+#[derive(Clone)]
+pub struct HostLayer {
+    listen: Vec<Multiaddr>,
+}
+
+impl HostLayer {
+    pub fn new(listen: Vec<Multiaddr>) -> Self {
+        Self { listen }
+    }
+}
+
+impl<S> Layer<S> for HostLayer {
+    type Service = HostMiddleware<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        HostMiddleware {
+            inner,
+            listen: self.listen.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct HostMiddleware<S> {
+    inner: S,
+    listen: Vec<Multiaddr>,
+}
+
+impl<S> Service<Request<Body>> for HostMiddleware<S>
+where
+    S: Service<Request<Body>, Response = Response<Body>, Error = Infallible> + Send + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = Infallible;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: Request<Body>) -> Self::Future {
+        let result = host(request.headers(), self.listen.clone());
+
+        if let Err(err) = result {
+            let error_response = err.into_response();
+            return Box::pin(async move { Ok(error_response) });
+        }
+
+        let future = self.inner.call(request);
+
+        Box::pin(async move {
+            let response: Response = future.await?;
+            Ok(response)
+        })
```

Comment:
```suggestion
        self.inner.call(request)
```

Reply:
updated to  Box::pin(self.inner.call(request)) as this is wanted return type

Reply:
Interesting, I expected `self.inner.call(request)` to return `Pin<Box<..>>`, weird..

Reply:
Ah, this makes sense now.. `<S as Service<..>>::Future` is not guaranteed to be a `Pin<Box<..>>`.